### DISCUSSION TITLE
Kubernetes와 Docker로 한 번에 끝내는 컨테이너 기반 MSA CH02.06 질문입니다

### DISCUSSION THREAD

**Discussion by rlaghtjr2** [tedilabs/fastcampus-devops#67] [Q&A] [answered]
안녕하세요 Kubernetes와 Docker로 한 번에 끝내는 컨테이너 기반 MSA 강의를 잘 듣고있습니다 감사합니다 :)
다름이 아니라 MAC OS로 진행하던중 ,network설정을 하면 connection refused 및 permission denied가 뜨는데, 혹시 다른 설정이 필요한지 알고싶습니다.

### host network
```
docker run -d --network=host grafana/grafana
```
이후 정상 running되는걸 확인하고
```
docker inspect [container]
``` 
을 통해 3000port 가 expose되어있는걸 확인하였습니다.
```
curl localhost:3000
```
진행시
```
curl: (7) Failed to connect to localhost port 3000: Connection refused
```
와 같은 에러가 발생합니다.

### bridge network
```
# network 생성
docker network create --driver=bridge fastcampus

# container 2개를 생성 후 네트워크로 엮어줌.
docker run -d --network=fastcampus --net-alias=hello nginx
docker run -d --network=fastcampus --net-alias=grafana grafana/grafana
```

이후 
``` 
docker ps
docker network ls
```
를 통해 container가 정상작동되고, network가 생성 잘 된것으로 확인되는데

```
docker exec -it [grafana container] bash
```
을 통해 접근 후
```
wget hello
```
입력시
```
wget: can't open 'index.html': Permission denied
```
에러가 발생합니다.

혹시 연결에 관해 다른 설정이 필요한것일까요 ?

**Comment by posquit0** [comment, marked as answer]
안녕하세요,
강의 잘 들어주셔서 감사합니다.

우선, 강의 처음에 소개한 내용처럼 도커 컨테이너 기술은 리눅스 운영체제의 기술입니다. Docker for Mac이나 Docker for Window 를 사용한다면 여러가지 기술적 제약사항을 가지고 있어 예상치 못한 결과를 얻을 수 있습니다.
떄문에, 리눅스 기반 실습환경(AWS EC2 혹은 가상머신, 직접 남는 PC에 리눅스 설치 등)을 구성하시고 실습을 진행하시는 것을 추천드립니다.

Docker for Mac에서 호스트 네트워킹은 알려진 이슈를 가지고 있으며, `bridge` 네트워크를 이용하면서 `-p` (publish)로 포트 포워딩하는 것은 지원합니다.

컨테이너 간의 연결 또한 이슈가 있을 수 있습니다. ㅠㅠ

### 참고자료

- https://docs.docker.com/desktop/mac/networking/
